### PR TITLE
Support for indexed properties

### LINKED ISSUE #325: Support indexed properties
Requested by user.

### PR DESCRIPTION
**Pull request by alazier**
@appden 

Closes #325

### CODE REVIEW

File: scripts/download-core.sh [hunk 1]
```diff
@@ -3,7 +3,7 @@
 set -e
 set -o pipefail
 
-: ${REALM_CORE_VERSION:=0.96.2} # set to "current" to always use the current build
+: ${REALM_CORE_VERSION:=0.97.0} # set to "current" to always use the current build
```
Comment (bdash) [inline scripts/download-core.sh]: You'll want to skip core v0.97.0. It crashes on 32-bit iOS devices. v0.97.1 and newer have the fix.

Comment (alazier) [inline scripts/download-core.sh]: Thanks for the info. We should update object store with the newer version as well (that's where I pulled this from).

File: src/object-store/src/property.hpp [hunk 2]
```diff
@@ -55,6 +55,7 @@ namespace realm {
 
         size_t table_column = -1;
         bool requires_index() const { return is_primary || is_indexed; }
+        bool is_indexable() const { return type == PropertyTypeInt || type == PropertyTypeBool || type == PropertyTypeString || type == PropertyTypeDate; }
```
Comment (appden) [inline src/object-store/src/property.hpp]: If the plan is to migrate dates to double fields, then perhaps we shouldn't let them be indexed?

File: scripts/prepublish.sh [hunk 3]
```diff
@@ -2,7 +2,8 @@
 
 set -e
 set -o pipefail
-
+export REALM_CORE_VERSION=`bash ./scripts/download-core.sh --version`
```
Comment (appden) [inline scripts/prepublish.sh]: Please use `$(...)` syntax and remove the `bash` part

File: scripts/prepublish.sh [hunk 4]
```diff
@@ -2,7 +2,8 @@
...
+echo "Core Version:$REALM_CORE_VERSION"
```
Comment (appden) [inline scripts/prepublish.sh]: Space after colon :smile:

File: react-native/android/build.gradle [hunk 5]
```diff
@@ -27,8 +27,8 @@ import org.apache.tools.ant.filters.ReplaceTokens
 // We download various C++ open-source dependencies into downloads.
 // We then copy both the downloaded code and our custom makefiles and headers into third-party-ndk.
 // After that we build native code from src/main/jni with module path pointing at third-party-ndk.
-ext.coreVersion = '0.96.2'
 
+ext.coreVersion = "$System.env.REALM_CORE_VERSION"
```
Comment (appden) [inline react-native/android/build.gradle]: Good idea, but we support calling gradle directly, so I would switch this to add an `npm run` script to `package.json` (like `get-core-version`, which calls `scripts/download-core.sh --version`) and run that just like the line below that sets `currentVersion`

File: tests/js/realm-tests.js [hunk 6]
```diff
@@ -258,6 +258,52 @@ module.exports = BaseTest.extend({
...
+
+        TestCase.assertThrows(function() {
+            IndexedTypes.properties = { dataCol: {type: 'data', indexed: true} }
+            new Realm({schema: [IndexedTypes], path: '4'});
+        });
+
+        // primary key
+        IndexedTypes.primaryKey = 'boolCol';
+        IndexedTypes.properties = { boolCol: {type: 'bool', indexed: true} }
+        new Realm({schema: [IndexedTypes], path: '5'});
```
Comment (appden) [inline tests/js/realm-tests.js]: Assuming this is to test that it does **not** throw an exception, so could you add that to the comment above

**Comment by appden** [comment]
Also, can you update the API docs

File: src/object-store/src/property.hpp (hunk 2 above)
Comment (alazier) [inline src/object-store/src/property.hpp]: You should join the object store repo so you can give this feedback to the commiter of this code.

Comment (appden) [inline src/object-store/src/property.hpp]: Haha ok, but if we foresee this changing sometime soon, then perhaps we should skip documenting it?

**Comment by appden** [comment]
Looks good. I'm still wary about indexing dates, but if that's not going to change anytime soon, then we can leave it. :+1:

**Pull request merged**